src/Npgsql/NpgsqlConnector.cs

```diff
@@ -832,25 +832,46 @@ void SetSocketOptions(Socket socket)
                 throw new ArgumentException("If TcpKeepAliveInterval is defined, TcpKeepAliveTime must be defined as well");
             if (Settings.TcpKeepAliveTime > 0)
             {
-                if (!PGUtil.IsWindows)
-                    throw new PlatformNotSupportedException(
-                        "Npgsql management of TCP keepalive is supported only on Windows. " +
-                        "TCP keepalives can still be used on other systems but are enabled via the TcpKeepAlive option or configured globally for the machine, see the relevant docs.");
-
-                var time = Settings.TcpKeepAliveTime;
-                var interval = Settings.TcpKeepAliveInterval > 0
+                var timeMilliseconds = Settings.TcpKeepAliveTime;
+                var intervalMilliseconds = Settings.TcpKeepAliveInterval > 0
                     ? Settings.TcpKeepAliveInterval
                     : Settings.TcpKeepAliveTime;
 
-                // For the following see https://msdn.microsoft.com/en-us/library/dd877220.aspx
-                var uintSize = Marshal.SizeOf(typeof(uint));
-                var inOptionValues = new byte[uintSize * 3];
-                BitConverter.GetBytes((uint)1).CopyTo(inOptionValues, 0);
-                BitConverter.GetBytes((uint)time).CopyTo(inOptionValues, uintSize);
-                BitConverter.GetBytes((uint)interval).CopyTo(inOptionValues, uintSize * 2);
-                var result = socket.IOControl(IOControlCode.KeepAliveValues, inOptionValues, null);
-                if (result != 0)
-                    throw new NpgsqlException($"Got non-zero value when trying to set TCP keepalive: {result}");
+
+                if (PGUtil.IsWindows)
+                {
+                    // Historically, we only supported setting TCP keepalive parameters on Windows, which
```

Comment:
I don't think millisecond granularity is a sufficiently good reason to keep using this Windows-specific code - we should ideally just switch to always using the new cross-platform .NET Core API. This code would still be used with legacy TFMs though.

We can mostly keep backwards compat by round the millisecond value, but I think we should consider just switching to seconds on the connection string parameter. This would obviously be a breaking change, but I think it makes sense to do it. Our other timeouts in the connection string are seconds, and high granularity for TCP keepalives really doesn't seem important.

What are your thoughts @chrisdcmoore @YohDeadfall?

Reply:
I'm fine with breaking changes, you know (: If you feel okay with them too, let's go this way.

Reply:
I'm OK with this breaking change - let's just change the connection string to seconds. We'll highlight this in our release notes, and it's unlikely too many people are using TCP keepalives anyway.

Reply:
A breaking change to switch to seconds everywhere works for me - I agree that millisecond granularity doesn't really make that much sense in the context of keepalives. 👍